KsApi/extensions/MockGraphQLClient.swift

```diff
@@ -0,0 +1,116 @@
+import Apollo
+import Foundation
+import ReactiveSwift
+
+class MockGraphQLClient: ApolloClientType {
+  // MARK: - Base Properties
+
+  var client: ApolloClient {
+    let url = URL(string: "https://kickstarter.com")!
+
+    return ApolloClient(url: url)
+  }
+
+  static let shared = MockGraphQLClient()
+
+  // MARK: Public functions
+
+  /// Placeholder implementation because protocol definition used in `Service`
+  public func fetch<Query: GraphQLQuery>(query _: Query) -> SignalProducer<Query.Data, ErrorEnvelope> {
+    .init(value: Query.Data(unsafeResultMap: [:]))
+  }
+
+  /// Placeholder implementation because protocol definition used in `Service`
+  public func perform<Mutation: GraphQLMutation>(
+    mutation _: Mutation
+  ) -> SignalProducer<Mutation.Data, ErrorEnvelope> {
+    .init(value: Mutation.Data(unsafeResultMap: [:]))
+  }
+}
+
+/** Implementation of optional `fetch` and `perform` with `result` useful for mocking data.
+ */
+extension ApolloClientType {
+  public func fetchWithResult<Query: GraphQLQuery, Data: Decodable>(
+    query _: Query,
+    result: Result<Data, ErrorEnvelope>?
+  ) -> SignalProducer<Data, ErrorEnvelope> {
+    producer(for: result)
+  }
+
+  public func performWithResult<Mutation: GraphQLMutation, Data: Decodable>(
+    mutation _: Mutation,
+    result: Result<Data, ErrorEnvelope>?
+  ) -> SignalProducer<Data, ErrorEnvelope> {
+    return producer(for: result)
+  }
+
+  public func dataFromProducer<Data: Decodable>(_ producer: SignalProducer<Data, ErrorEnvelope>) -> Data? {
```

Comment:
A bit confused why we're handling data and error in two separate methods? Shouldn't we model this similarly to how we handle this in the `from(data:) -> T` and `producer(from data:)` methods?

Reply:
I split them up because they return two different types, one an `ErrorEnvelope?` and the other `Data?`.
I renamed them to be more readable and closer to our producer functions, but left them in two methods just for clarity because we call one just for data and the other specifically when we expect an error.
Could have made them return a tuple in one function, but again readability is helpful here.